proof/refine/ARM/Finalise_R.thy

```diff
@@ -4362,62 +4489,388 @@ global_interpretation delete_one
+                and pspace_aligned and pspace_distinct and valid_objs
+                and (\<lambda>s. \<exists>n. ko_at (Structures_A.SchedContext x n) scPtr s)"
+              in stronger_corres_guard_imp)
+         apply (rule corres_guard_imp)
+           apply (rule_tac F="(sc_replies x \<noteq> []) = (\<exists>y. scReply sc = Some y)" in corres_gen_asm2)
+           apply (rule corres_when)
+            apply clarsimp
+           apply (rule_tac F="scReply sc = Some (hd (sc_replies x))" in corres_gen_asm2)
+           apply clarsimp
+           apply (rule corres_split_new[OF replyNext_update_corres_empty])
+             apply (rule update_sc_reply_stack_update_ko_at'_corres)
+            apply wpsimp+
+          apply (clarsimp simp: obj_at_def)
+          apply (frule (1) valid_sched_context_objsI)
+          apply (clarsimp simp: valid_sched_context_def list_all_def obj_at_def)
+         apply clarsimp
+         apply (case_tac "sc_replies x"; simp)
+        apply assumption
+       apply (clarsimp simp: obj_at_def  is_sc_obj)
+       apply (frule state_relation_sc_replies_relation)
+       apply (subgoal_tac "scReply sc = hd_opt (sc_replies sca)")
+        apply (intro conjI)
+          apply clarsimp
+         apply clarsimp
+         apply (erule (1) reply_at_cross[rotated])
+          apply (frule (1) valid_sched_context_objsI)
+          apply (clarsimp simp: valid_sched_context_def list_all_def obj_at_def)
+         apply fastforce
+        apply (erule (1) sc_replies_relation_prevs_list)
+        apply (clarsimp simp: obj_at'_real_def ko_wp_at'_def projectKO_sc)
+       apply (erule (1) sc_replies_relation_scReply[symmetric])
+       apply (clarsimp simp: obj_at'_real_def  ko_wp_at'_def projectKO_sc)
+      apply wpsimp+
+    apply (fastforce simp: sym_refs_asrt_def)+
+  done
+
+lemma schedContextUnbindYieldFrom_corres:
+  "corres dc (einvs and sc_at scPtr and K (scPtr \<noteq> idle_sc_ptr)) (invs' and sc_at' scPtr)
+          (sched_context_unbind_yield_from scPtr) (schedContextUnbindYieldFrom scPtr)"
+  apply (clarsimp simp: sched_context_unbind_yield_from_def schedContextUnbindYieldFrom_def
+                        maybeM_when)
+  apply add_sym_refs
+  apply (rule corres_stateAssert_implied[where P'=\<top>, simplified])
+   apply (rule corres_guard_imp)
+     apply (rule corres_split_new[OF get_sc_corres])
+       apply (case_tac sca)
+       apply (clarsimp simp: sc_relation_def)
+       apply (wpfix add: sched_context.sel)
+       apply (rule corres_when)
+        apply (clarsimp simp: sc_relation_def)
+       apply (rule complete_yield_to_corres)
+      apply wpsimp+
+    apply (fastforce dest!: invs_valid_objs valid_objs_ko_at
+                      simp: valid_obj_def valid_sched_context_def)
+   apply (fastforce dest!: sc_ko_at_valid_objs_valid_sc'
+                     simp: valid_obj'_def valid_sched_context'_def )
+  apply (clarsimp simp: sym_refs_asrt_def)
+  done
+
+lemma set_sc_obj_ref_twice_rewrite:
```

Comment:
I've got a more general version of this, if you'd like (well, I resurrected it from when `refill_new` and `refill_update` were defined in terms of one big `update_sched_context. I think Matt wrote this proof.)

```
lemma update_sched_context_decompose:
   "update_sched_context scp (λsc. f (g sc))
    = (do update_sched_context scp g; update_sched_context scp f od)"
  apply (rule ext)
  by (clarsimp simp: update_sched_context_def get_object_def set_object_def a_type_simps
                     gets_def get_def put_def return_def fail_def assert_def bind_def
              split: Structures_A.kernel_object.splits)
```

Reply:
Thanks, that's a much nicer rule (and proof)! Any suggestions for where it should go?

Reply:
If you'd like to put it in AInvs, then I reckon KHeap is as good as any. If that seems too high up, then maybe one of the DetSched files, probably no later than the top part of DetSchedSchedule (before it gets gnarly)